core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/hashJoin/HashJoinOpExec.scala

```diff
@@ -31,66 +34,176 @@ class HashJoinOpExec[K](
-
-                if (attributeName != probeAttributeName) {
-                  builder.add(
-                    new Attribute(
-                      if (buildSchema.getAttributeNames.contains(attributeName))
-                        attributeName + "#@1"
-                      else attributeName,
-                      attribute.getType
-                    ),
-                    t.getFields.get(i)
-                  )
-                }
-              }
-              builder.build()
-            })
-            .toIterator
         }
       case Right(_) =>
-        if (input == buildTable) {
+        if (input == buildTable && !isBuildTableFinished) {
+          // the first input is exhausted, building phase finished
           isBuildTableFinished = true
+          Iterator()
+        } else {
+          // the second input is exhausted, probing phase finished
+          if (leftOuterJoin) {
+            performLeftOuterJoin
+          } else {
+            Iterator()
+          }
         }
-        Iterator()
+    }
+  }
+
+  private def performLeftOuterJoin: Iterator[Tuple] = {
+    buildTableHashMap.valuesIterator
+      .filter({ case (_: ArrayBuffer[Tuple], joined: Boolean) => !joined })
+      .flatMap {
+        case (tuples: ArrayBuffer[Tuple], _: Boolean) =>
+          tuples
+            .map((tuple: Tuple) => {
+              // creates a builder
+              val builder = Tuple.newBuilder(operatorSchemaInfo.outputSchema)
+
+              // fill the probe tuple attributes as null, since no match
+              fillNonJoinFields(
+                builder,
+                probeSchema,
+                Array.fill(probeSchema.getAttributesScala.length)(null),
+                resolveDuplicateName = true
+              )
 
+              // fill the build tuple
+              fillNonJoinFields(builder, buildSchema, tuple.getFields.toArray())
+
+              // fill the join attribute (align with build)
+              builder.add(
+                buildSchema.getAttribute(buildAttributeName),
+                tuple.getField(buildAttributeName)
+              )
+
+              // build the new tuple
+              builder.build()
+            })
+            .toIterator
+      }
+  }
+
+  private def performJoin(probeTuple: Tuple, matchedTuples: ArrayBuffer[Tuple]): Iterator[Tuple] = {
+
+    matchedTuples
+      .map(buildTuple => {
+        // creates a builder with the build tuple filled
+        val builder = Tuple
+          .newBuilder(operatorSchemaInfo.outputSchema)
+          .add(buildTuple)
+
+        // append the probe tuple
+        fillNonJoinFields(
+          builder,
+          probeSchema,
+          probeTuple.getFields.toArray(),
+          resolveDuplicateName = true
+        )
+
+        // build the new tuple
+        builder.build()
+      })
+      .toIterator
+  }
+
+  private def performRightOuterJoin(tuple: Tuple): Iterator[Tuple] = {
```

Comment:
change to change to `performRightAntiJoin`

Reply:
done